File: resources/EventLogger.js
// *********************
// Author: Tauvic Ritter (https://github.com/Tauvic)
// Description: Event Logger for Robot Framework
// License: MIT License
// *********************

//const { BrowserContext, Page } = require('playwright');

/**
 * @typedef {object} RequestData
 * @property {number} requestID
 * @property {string} method
 * @property {string} resourceType
 * @property {string} url
 * @property {string | null} postData
 * @property {string | null} failure
 */

/**
 * @typedef {object} ResponseData
 * @property {number} status
 * @property {string} statusText
 * @property {boolean} ok
 * @property {string | null} content
 */

/**
 * @typedef {object} EventLogEntry
 * @property {Number} time
 * @property {string} context
 * @property {string} event
 * @property {string} type
 * @property {any} data
 */

/**
 * @typedef {object} ContextData
 * @property {string | null} alerts
 * @property {number} requestID
 * @property {Set<number>} activeRequests
 * @property {number} lastEventTime
 * @property {number} maxWait
 * @property {number} minIdle
 * @property {Promise<{resolve: (value?: any) => void, reject: (reason?: any) => void} | null>} waitPromise
 * @property {NodeJS.Timeout | null} waitIdle
 * @property {NodeJS.Timeout | null} waitTimeout
 * @property {string | null} context
 * @property {EventLogEntry[]} events
 */

/**
 * Initializes the event logger.
 * @param {number} [maxWait=10000] - The maximum wait time for events.
 * @param {number} [minIdle=150] - The minimum idle time for network requests.
 * @param {string | null} [alerts=null] - The XPath for alert elements.
 * @param {BrowserContext} context - The Playwright BrowserContext.
 * @param {function} logger - The logger function.
 * @returns {Promise<void>}
 */
async function initEventLogger(
    maxWait = 10000,
    minIdle = 150,
    alerts = null,
    context,
    logger
) {
    // Initialize Event Logger
    // Call directly after browser context is created
    // This function only refers to the browser context
    // add event handlers to log API and UI events:
    //   API requests and responses
    //   API request failures
    //  Dom changes (used for alerts, warnings, info messages, toast messages)
    //  Console messages

    logger(`init Event Logger start`);

    if ('cfg' in context) {
        //Prevent running this twice
        //We assume that the context is not reused in the same test
        logger(`Init API Requests called twice !! `);
        throw new Error("Init API Requests called twice !!");
    }

    // create data structure to store results in context
    /** @type {ContextData} */
    context['cfg'] = {
        // alert handling
        alerts: alerts,             // alerts to check for
        // request/response handling
        requestID: 0,               // request ID
        activeRequests: new Set(),  // active requests to check for
        lastEventTime: 0,           // last time a request/response was received
        maxWait: maxWait,           // maximum wait time to check for events
        minIdle: minIdle,           // minimum idle time to check for events
        waitPromise: null,
        waitIdle: 0,                // id of idle timer
        waitTimeout: 0,             // id of timeout timer  
        events: []                  // array to store events  
    };

    context.addInitScript(() => {

        // This script runs on a page in the browser, not in the Node.js context, so context does not work here
        // We use MutationObserver to monitor DOM changes and classify them as alerts or toasts

        alerts = {};
        alertID = 0;
        alertTimer = null;

        function isGenerallyVisible(element) {
            const style = window.getComputedStyle(element)
            return style.display !== 'none' && style.visibility !== 'hidden' && style.opacity !== '0'
        }

        function isVisibleInViewport(element) {
            const rect = element.getBoundingClientRect()
            return (
                rect.top >= 0 &&
                rect.left >= 0 &&
                rect.bottom <= (window.innerHeight || document.documentElement.clientHeight) &&
                rect.right <= (window.innerWidth || document.documentElement.clientWidth)
            )
        }    
        
        function isElementFullyVisible(element) {
            return isGenerallyVisible(element) && isVisibleInViewport(element) && element.checkVisibility()
        }

        function updateStatus(alertID) {       

            const alertData = alerts[alertID];
            const wasVisible = alertData.visible;
            const isVisible = isElementFullyVisible(alertData.node);

            if (wasVisible == true) {
                const delta = Date.now()-alertData.updated;
                alertData.shown += delta; // update the shown time
            }

            alertData.visible = isVisible; // update the visibility status
            alertData.updated = Date.now(); // update the last updated time

            if (isVisible !== wasVisible) console.debug(`Alert: Updated`,alertData);
        }

        function classify(node) {

            //TO-DO: check for aria role

            const className = node.className;
            // find the word alert or toast
            const isAlert = /\balert\b/;
            const alertType = /\balert-(success|info|warning|error|danger)\b/g
            const isToast = /\b(toast|ngx-toastr)\b/;
            const toastType = /\btoast-(success|info|warning|error|danger)\b/g
            
            let   msgType = null;                    
            let   msgClass= null;
            let   msgTitle= null;
            let   msgText = null;    
                
            if (isAlert.test(className)) {
                const match = className.match(alertType);
                if (match && match.length===1) {
                    msgType  = 'alert'
                    msgClass = match[0]
                    msgText  = node.innerText.trim()
                }                  
            } else if (isToast.test(className)) {
                const match = className.match(toastType)
                if (match && match.length===1) {
                    msgType  = 'toast'
                    msgClass = match[0]
                    const titleNode = node.querySelector(".toast-title");
                    if (titleNode) {
                    msgTitle = titleNode.innerText.trim();
                    const msgNode = node.querySelector(".toast-message");
                    msgText = msgNode.innerText.trim();

                    } else
                    msgText  = node.innerText.trim()
                }      
            }

            if (!msgType) {
                node.dataset.alertID = null; // classify as not an alert
                return;
            }

            const newID = ++alertID;
            node.dataset.alertID = newID; // store the alert ID in the node
            
            const alertData = {alert: newID, 
                            class: msgClass,                            
                            visible:isElementFullyVisible(node),
                            shown: 0,
                            text: msgText,                                    
                            title: msgTitle,                           
                            type: msgType,
                            node:node,
                            updated: Date.now()};

            alerts[newID] = alertData; // store the alert in the alerts array

            console.debug('Alert: Created',alertData);

            if (Object.keys(alerts).length === 1) {
                alertTimer = setInterval(() => {

                    for (const [alertID, alertData] of Object.entries(alerts)) {
                        if (document.body.contains(alertData.node)) {
                            updateStatus(alertID);
                        } else {
                            delete alerts[alertID]; // remove the alert from the alerts array
                            console.debug(`Alert: Removed left=[${Object.keys(alerts)}]`,alertData);
                        }
                    };

                    if (Object.keys(alerts).length === 0) {
                        clearInterval(alertTimer); // stop the timer if there are no alerts
                        alertTimer = null;
                        console.debug(`Alert: Cleared all alerts`);
                    }
                }, 50); // check every 50ms
            }
        }

        function checkNode(node) {

            // check if node is element and has className
            if (!node || node.nodeType !== Node.ELEMENT_NODE || !node.className) return;                 

            const alertID = node.dataset?.alertID

            // alertID is set when the node is classified as an alert
            if (alertID > 0) {
                updateStatus(alertID);
            } else if (alertID === undefined) {
                classify(node); // classify the node as an alert or not set the alertID to null
            }

        }        

        // create a new instance of `MutationObserver` named `observer`,
        const observer = new MutationObserver((mutations) => {

            // for each mutation in the array of mutations passed to the callback function
            for (const mutation of mutations) {

                checkNode(mutation.target);

                switch(mutation.type) {
                    case 'childList':
                        for (const node of mutation.addedNodes) {
                            checkNode(node);
                        };
                        for (const node of mutation.removedNodes) {
                            checkNode(node);
                        };
                        break;   
                    case "attributes":
                        break;
                    case "characterData":                     
                        break;
                }
            }

        });

        // call `observe()`, passing it the element to observe, and the options object
        observer.observe(window.document, {
            subtree: true,
            childList: true
        });

    });

    // listen for requests
    context.on('request', request => {
        /** @type {ContextData} */
        const cfg = context.cfg;
        const url = new URL(request.url());

        if (['xhr','fetch'].includes(request.resourceType()) && url.hostname.includes(cfg.hostname)) {

            cfg.requestID += 1;
            request.requestID = cfg.requestID;
            cfg.activeRequests.add(request.requestID);
            cfg.lastEventTime = Date.now();

            const methodsWithBody = ['POST', 'PUT', 'PATCH'];
            const requestMethod = request.method();
            let content = null;

            // Check if the request method is one of the methods that can have a body
            if (methodsWithBody.includes(requestMethod)) {
                const contentType = request.headers()['content-type'];
                if (contentType && contentType.includes('application/json')) {
                  // Attempt to parse the post data as JSON to confirm
                  content = request.postDataJSON();
                  if (content) content = JSON.stringify(content, null, 2);
                } else {
                  // If not JSON, use the raw post data
                  content = request.postData();
                }
            }

            /** @type {RequestData} */
            const rqd = {
                requestID: request.requestID,
                method: requestMethod,
                resourceType: request.resourceType(),
                url: request.url(),
                postData: content,
                failure: null,
            }

            cfg.events.push({ time: Date.now(), event: 'request', 'type': 'INFO', data: rqd });
            //cfg.events.push({ 'time': Date.now(), 'event': 'console', 'type': 'DEBUG', 'data': `waitIdle size=${cfg.activeRequests.size} promise=${cfg.waitPromise} id=${cfg.waitIdle}` });
            if (cfg.waitIdle) {
                //cfg.events.push({ 'time': Date.now(), 'event': 'console', 'type': 'DEBUG', 'data': `clear waitIdle id=${cfg.waitIdle}` });
                clearTimeout(context.cfg.waitIdle);
                cfg.waitIdle = null;
            }
        } 
    });

    // listen for requests failed
    context.on('requestfailed', request => {

        /** @type {ContextData} */
        const cfg = context.cfg;

        if (cfg.activeRequests.has(request.requestID)) {

            cfg.activeRequests.delete(request.requestID);
            cfg.lastEventTime = Date.now();

            const methodsWithBody = ['POST', 'PUT', 'PATCH'];
            const requestMethod = request.method();
            let content = null;

            // Check if the request method is one of the methods that can have a body
            if (methodsWithBody.includes(requestMethod)) {
                const contentType = request.headers()['content-type'];
                if (contentType && contentType.includes('application/json')) {
                  // Attempt to parse the post data as JSON to confirm
                  content = request.postDataJSON();
                  if (content) content = JSON.stringify(content, null, 2);
                } else {
                  // If not JSON, use the raw post data
                  content = request.postData();
                }
            }

            /** @type {RequestData} */
            const rqd = {
                requestID: request.requestID,
                method: requestMethod,
                resourceType: request.resourceType(),
                url: request.url(),
                postData: content,
                failure: request.failure().errorText,
            }

            cfg.events.push({ time: Date.now(), event: 'request', type: 'WARN', data: rqd });
            //cfg.events.push({ 'time': Date.now(), 'event': 'console', 'type': 'DEBUG', 'data': `waitIdle size=${cfg.activeRequests.size} promise=${cfg.waitPromise}` });
            if (cfg.activeRequests.size === 0 && context.cfg.waitPromise) {
                cfg.waitIdle = setTimeout(cfg.waitPromise.resolve, cfg.minIdle);
                //cfg.events.push({ 'time': Date.now(), 'event': 'console', 'type': 'DEBUG', 'data': `set waitIdle id=${cfg.waitIdle} time=${cfg.minIdle}`  });
            }
        }
    });

    // listen for responses finished
    context.on('requestfinished', async request => {

        /** @type {ContextData} */
        const cfg = context.cfg;

        if (cfg.activeRequests.has(request.requestID)) {

            cfg.activeRequests.delete(request.requestID);
            cfg.lastEventTime = Date.now();

            const methodsWithBody = ['POST', 'PUT', 'PATCH'];
            const requestMethod = request.method();
            let content = null;

            // Check if the request method is one of the methods that can have a body
            if (methodsWithBody.includes(requestMethod)) {
                const contentType = request.headers()['content-type'];
                if (contentType && contentType.includes('application/json')) {
                  // Attempt to parse the post data as JSON to confirm
                  content = request.postDataJSON();
                  if (content) content = JSON.stringify(content, null, 2);
                } else {
                  // If not JSON, use the raw post data
                  content = request.postData();
                }
            } 

            /** @type {RequestData} */
            const rqd = {
                requestID: request.requestID,
                method: requestMethod,
                resourceType: request.resourceType(),
                url: request.url(),
                postData: content,
                failure: null
            }

            const resp = await request.response();
            const contentType = resp.headers()['content-type'];
            content = null;

            try {
                if (contentType && contentType.includes('application/json')) {
                    content = await resp.json();
                    content = JSON.stringify(content, null, 2);
                } else {
                    // If not JSON, use the raw post data
                    content = await resp.text();    
                }    
            } catch (error) {
                cfg.events.push({ time: Date.now(), event: 'console', type: 'ERROR', data: error.toString() });
            }

            /** @type {ResponseData} */
            const rsd = {
                status: resp.status(),
                statusText: resp.statusText(),
                ok: resp.ok(),
                content: content
            }
            cfg.events.push({ time: Date.now(), event: 'response', type: 'INFO', data: [rqd,rsd] });
            //cfg.events.push({ 'time': Date.now(), 'event': 'console', 'type': 'DEBUG', 'data': `waitIdle size=${cfg.activeRequests.size} promise=${cfg.waitPromise}` });
            if (cfg.activeRequests.size === 0 && context.cfg.waitPromise) {
                cfg.waitIdle = setTimeout(cfg.waitPromise.resolve, cfg.minIdle);
                //cfg.events.push({ 'time': Date.now(), 'event': 'console', 'type': 'DEBUG', 'data': `set waitIdle id=${cfg.waitIdle} time=${cfg.minIdle}`  });
            }

        }
    });

    // listen for console messages
    context.on('console', async (message) => {
        /** @type {ContextData} */
        const cfg = context.cfg;

        if (message.text() != "JSHandle@error") {
            cfg.events.push({ time: Date.now(), event: 'console', type: message.type().toUpperCase(), data: message.text() });
            return;
        };

        const messages = await Promise.all(message.args().map((arg) => {
            return arg.getProperty("message");
        }));

        msg = `${messages}`
        cfg.events.push({ time: Date.now(), event: 'console', type: message.type().toUpperCase(), data: msg });
    });

    context.on('page', page => {

        page.on('framenavigated', async (frame) => {
            const url = page.url();
            // save hostname to context: used for checking api requests
            context.cfg.hostname = await page.evaluate(()=> {return location.hostname});;
            context.cfg.events.push({ time: Date.now(), event: 'console', type: 'INFO', data: `Navigated: url=${url}`, context: url });
        });

        page.on('pageerror', data => {
            context.cfg.events.push({ time: Date.now(), event: 'console', type: 'ERROR', data: data.message });
        });
    });

    logger(`init Event Logger finish`);
}

async function checkAlerts(page, path, timeout = 100) {

    //context.cfg.events.push({ 'time': Date.now(), 'event': 'console', 'type': 'DEBUG', 'data': 'check Alerts' });

    try {
        await page.locator(path).waitFor({ state: 'visible', timeout: timeout });
    } catch (error) {
        return; // Ignore timeout errors
    }

    const elements = await page.locator(path).all();
    if (elements && elements.length > 0) {
        const texts = [];
        for (const element of elements) {
            texts.push(await element.innerText());
        }
        throw new Error(texts.join(", "));
    }
}

/**
 * Wait for all active requests to finish and for the minimum idle time to pass.
 * @param {ContextData} cfg
 */
function waitPromise(cfg) {

    return new Promise((resolve, reject) => {

        // share this promise with the request handler
        // so that we can resolve it when all requests are done
        cfg.waitPromise = { resolve, reject };

        const waiting = cfg.activeRequests.size;      // Number of requests in flight
        const idle = Date.now() - cfg.lastEventTime;  // Time since last event

        //cfg.events.push({ 'time': Date.now(), 'event': 'console', 'type': 'DEBUG', 'data': `wait Started waiting=${waiting} idle=${idle}` });

        if (waiting === 0) {
            if (idle > cfg.minIdle) {
                //cfg.events.push({ 'time': Date.now(), 'event': 'console', 'type': 'DEBUG', 'data': `wait Resolve` });
                resolve();
                return;
            }
            else {
                cfg.waitIdle = setTimeout(resolve, cfg.minIdle - idle);
                //cfg.events.push({ 'time': Date.now(), 'event': 'console', 'type': 'DEBUG', 'data': `wait set id=${cfg.waitIdle} time=${cfg.minIdle - idle}` });
            }
        }

        cfg.waitTimeout = setTimeout(() => {
            reject(new Error(`waitForApiEvents: Max wait time of ${cfg.maxWait}ms exceeded outstanding=${[...cfg.activeRequests].join(',')} id=${cfg.waitTimeout}`));
        }, cfg.maxWait);
        //cfg.events.push({ 'time': Date.now(), 'event': 'console', 'type': 'DEBUG', 'data': `set waitTimeout id=${cfg.waitTimeout} time=${cfg.maxWait}` });

    });

}

// Wait for all outstanding events
async function waitForEvents(context, page) {

    const cfg = context.cfg;
    if (!cfg) return;
    const startTime = Date.now();

    // Wait for all events that we expect to happen
    try {
        // Check for alerts first because no need to wait for API events if we fail on alerts
        if (cfg.alerts) await checkAlerts(page, path = cfg.alerts);

        const startPromise = Date.now()

        await waitPromise(cfg);

        cfg.waitPromise = null;

        if (cfg.waitIdle) {
            //cfg.events.push({ 'time': Date.now(), 'event': 'console', 'type': 'DEBUG', 'data': `clear waitIdle id=${cfg.waitIdle}` });
            clearTimeout(cfg.waitIdle);
            cfg.waitIdle = null;
        }

        if (cfg.waitTimeout) {
            //cfg.events.push({ 'time': Date.now(), 'event': 'console', 'type': 'DEBUG', 'data': `clear waitTimeout id=${cfg.waitTimeout}` });
            clearTimeout(cfg.waitTimeout);
            cfg.waitTimeout = null;
        }

        //cfg.events.push({ 'time': Date.now(), 'event': 'console', 'type': 'DEBUG', 'data': `wait End delta=${Date.now()-startPromise}` });
        // Check for alerts again after all API responses have been processed
        if (cfg.alerts) await checkAlerts(page, path = cfg.alerts);
    } catch (error) {
        // usually an alert or timeout error
        cfg.events.push({ time: Date.now(), event: 'console', type: 'ERROR', data: error });
        throw error;
    }

    const endTime = Date.now();
    const delta = endTime - startTime;

    cfg.events.push({ time: endTime, event: 'console', type: 'DEBUG', data: `Wait satisfied in ${delta} ms` });
}

async function reportEvents(context) {

    if (!('cfg' in context)) return;

    return context.cfg.events;
}

initEventLogger.rfdoc = `
Init Event Logger https://api.practicesoftwaretesting.com
.`

waitForEvents.rfdoc = `
Wait for Events.
.`

reportEvents.rfdoc = `
Report Events.
.`

exports.__esModule = true;
exports.initEventLogger = initEventLogger;
exports.waitForEvents = waitForEvents;
exports.reportEvents = reportEvents;

File: resources/EventLogger.py
from robot.api.deco import library, keyword
from robot.api import logger
from robot import running, result
from robot.libraries.BuiltIn import BuiltIn
import datetime, time
from enum import Enum

import html

class LogLevel(Enum):
    DEBUG = 1
    INFO = 2
    WARN = 3
    ERROR = 4
    NONE = 5

@library(scope='TEST', version='0.0.1',listener='SELF')
class EventLogger:

    """
    Author: Tauvic Ritter (https://github.com/Tauvic)
    Description: Event Logger for Robot Framework
    License: MIT License
    """

    ROBOT_LISTENER_API_VERSION = 3

    def __init__(self):
      """
      Init Event Logger
      """
      logger.info(f"Event Listener init", also_console=True)
      self.waitAfter = None
      self.ommit = ['BuiltIn','Collections']
      self.level = 0
      self.local_events = []  # Local list to store events
      self.logLevel:LogLevel = LogLevel.INFO

    @keyword('Init')
    def initEventLogging(self, 
                         maxWait:int=10000,
                         minIdle:int=150,
                         waitAfter:str=None,
                         alerts:str=None,
                         logLevel:LogLevel=LogLevel.INFO):
        """
        Init Event Logging        
        """
        self.waitAfter = waitAfter
        self.logLevel  = logLevel
        BuiltIn().run_keyword('Browser.initEventLogger',maxWait,minIdle,alerts) 

    @keyword('Wait For Events')
    def waitForEvents(self):
        """
        Wait for Events
        """
        BuiltIn().run_keyword('Browser.waitForEvents') 
   
    @keyword('Report Event Logging')
    def report_event_logging(self,addToTestMessage=True) -> str:
        """
        Report Event Logging
        """
        # Retrieve all events from JavaScript
        js_events = BuiltIn().run_keyword('Browser.reportEvents')

        # Merge local events with JavaScript events based on timestamps
        all_events = self.local_events + js_events
        all_events.sort(key=lambda ev: ev['time'])
        
        # Generate HTML report
        report = self._generate_html_report(all_events).replace('{','\{')

        # Attach report to Robot Framework logs
        if report and addToTestMessage:
            args = {'append': True}
            BuiltIn().run_keyword('Set Test Message', f"*HTML*{report}", args)

        return report

    def _format_time(self,milliseconds):
      """Converts milliseconds since the epoch to HH:MM:SS.mmm format."""
      seconds = milliseconds / 1000
      dt_object = datetime.datetime.fromtimestamp(seconds)
      return dt_object.strftime("%H:%M:%S.%f")[:-3]

    def _generate_html_report(self, events):
        """
        Generate an HTML report from the merged events.
        """
        if not events: return ""

        level = 0
        context = None
        bgcolors = ['#5dade2','#1abc9c']  #  background colors
        bgcolor = 0 # Start with the first color
        html = '<table style="width:100%"><tr><th style="width:80px">Time</th><th style="width:60px">Source</th><th style="width:60px">Type</th><th style="width:30px">CTX</th><th style="width:100%">Data</th></tr>'
        for ev in events:            
            time = self._format_time(ev['time'])
            event = ev['event']
            event_type = ev['type']
            if LogLevel[event_type].value < self.logLevel.value: continue            
            data = self._format_event_data(ev)
            level = ev.get('level', level)

            if event == 'script':
               bonus = 0
            else:
               bonus = 1

            # Use margin-left for indentation
            indent_style = f"margin-left: {(level + bonus) * 20}px;"  # 20px per level

            new_context = ev.get('context', context)
            if new_context != context:
              bgcolor = 0 if bgcolor == 1 else 1  # Alternate background color
              context = new_context
              html += f"<tr style='background:{bgcolors[bgcolor]}'><td>{time}</td><td>{event}</td><td>{event_type}</td><td></td><td><div style='{indent_style}'>{data}</div></td></tr>"
            else:
              html += f"<tr><td>{time}</td><td>{event}</td><td>{event_type}</td><td style='background:{bgcolors[bgcolor]}'></td><td><div style='{indent_style}'>{data}</div></td></tr>"
           
        html += '</table>'
        return f"<details><summary>Event log:</summary>{html}</details>"

    def _format_event_data(self, event):
        """
        Format event data based on its type.
        """
        if event['event'] == 'request':
            rqd = event['data']
            rq_status = f"<span style='color:red'>{rqd['failure']}</span>" if rqd.get('failure') else "<span style='color:green'>Ok</span>"
            header = f"{rqd['requestID']:03d} {rqd['method']} {rqd['resourceType']} {rqd['url']} {rq_status}"
            details = f"<details><summary>{header}</summary><pre>{rqd.get('postData', '')}</pre></details>" if rqd.get('postData') else header            
            return details

        if event['event'] == 'response':
            rqd, rsd = event['data']
            s_c = 'green' if rsd['ok'] else 'red'
            header = f"{rqd['requestID']:03d} {rqd['method']} {rqd['resourceType']} {rqd['url']} <span style='color:{s_c}'>status={rsd['status']} {rsd['statusText']}</span>"
            details = f"<details><summary>{header}</summary><pre>{rsd.get('content', '')}</pre></details>" if rsd.get('content') else header
            return details

        if event['event'] == 'console':
            msg = event['data']
            if event['type'] == 'ERROR':
                msg = f"<span style='color:red'>{msg}</span>"
            return msg

        if event['event'] == 'script':
            return f"<span style='color:blue'>{event['data']}</span>"

        # Default formatting for other event types
        return str(event['data'])
      
    def start_user_keyword(self, kw:running.Keyword, impl, result:result.Keyword): 

      if result.status in ['FAIL','SKIP','NOT RUN']: return      

      try:
        args = " ".join([str(arg) for arg in result.args])    
        logger.info(f"Start: {kw.name} {args}",also_console=True)
        if  len(args) > 80:
          args = f'<details><summary>arguments</summary>{args}</details>'

        self.local_events.append({'time': int(time.time() * 1000), 
                                  'level': self.level,
                                  'event': 'script', 
                                  'type': 'INFO', 
                                  'data': f"Start: {kw.name} {args}"})
        self.level +=1

      except  Exception as ex:
        logger.warning(ex)        

    def end_user_keyword(self, kw:running.Keyword, impl, result:result.Keyword):

      if result.status in ['SKIP','NOT RUN']: return result

      try:
        msg = f"End : {kw.name} library={result.libname} status={result.status}"     
        logger.info(msg,also_console=True)   
        if (self.level > 0): self.level -=1           
        self.local_events.append({'time': int(time.time() * 1000), 
                                  'level': self.level,
                                  'event': 'script', 
                                  'type': 'INFO', 
                                  'data': msg})               
      except  Exception as ex:
        logger.warning(ex)                 

      # What library key words should we log and how
      # The keyword that we mention in waitAfter will be logged with start/end
      # the other keywords will be logged with start

    def start_library_keyword(self, kw:running.Keyword, impl, result:result.Keyword):
      
      if result.status in ['FAIL','SKIP','NOT RUN']: return

      if self.waitAfter and f"{result.libname}.{kw.name}" in self.waitAfter:  
        logPrefix = 'Start: '   
      else:
        if result.libname in self.ommit: return
        logPrefix = ''
      
      args = " ".join([str(arg) for arg in result.args])
      msg = f"{logPrefix}{kw.name} {args}"  
      logger.info(msg,also_console=True)

      if kw.name == 'Browser.waitForEvents': return       
 
      if  len(args) > 80:
        args = f'<details><summary>arguments</summary>{args}</details>'     
      
      msg = f"{logPrefix}{kw.name} {args}" 
      self.local_events.append({'time': int(time.time() * 1000), 
                                'level': self.level,
                                'event': 'script', 
                                'type': 'INFO', 
                                'data': msg})
    
      if logPrefix == 'Start: ':
        self.level +=1

    def end_library_keyword(self, kw:running.Keyword, impl, result:result.Keyword):

      if result.status in ['SKIP','NOT RUN']: return       
      
      # Auto start wait for Events
      if self.waitAfter and f"{result.libname}.{kw.name}" in self.waitAfter:    
        try:
          BuiltIn().run_keyword('Browser.waitForEvents') 
        except  Exception as ex:
          # logger.error(ex)
          result.status = 'FAIL'
          result.message = str(ex)
      else:
         return

      self.level -=1   
      msg = f"End:  {kw.name} library={result.libname} status={result.status}"  
      logger.info(msg,also_console=True)   
      self.local_events.append({'time': int(time.time() * 1000), 
                                'level': self.level,
                                'event': 'script', 
                                'type': 'INFO', 
                                'data': msg})


    def start_test(self, data: running.TestCase, result):
        if data.tags.match('ev:*'):
           for tag in data.tags:
              if tag.startswith('ev:ommit'):
                 self.ommit = tag.split('ev:ommit:')[1].split(',')
                 

  
